planner/core/plan.go

```diff
@@ -75,6 +75,42 @@ func enforceProperty(p *property.PhysicalProperty, tsk task, ctx sessionctx.Cont
 	return sort.attach2Task(tsk)
 }
 
+// optimizeByPartition insert `PhysicalPartition` to optimize performance by running in a parallel manner.
+func optimizeByPartition(pp PhysicalPlan, tsk task, ctx sessionctx.Context) task {
+	if tsk.plan() == nil {
+		return tsk
+	}
+
+	// Don't use `tsk.plan()` here, which will probably be different from `pp`.
+	// Eg., when `pp` is `NominalSort`, `tsk.plan()` would be its child.
+	switch p := pp.(type) {
+	case *PhysicalWindow:
+		if part := optimizeByPartition4Window(p, ctx); part != nil {
+			return part.attach2Task(tsk)
+		}
+	}
+	return tsk
+}
+
+func optimizeByPartition4Window(pp *PhysicalWindow, ctx sessionctx.Context) *PhysicalPartition {
+	concurrency := ctx.GetSessionVars().WindowConcurrency
+	if concurrency > 1 {
+		var tail, dataSource PhysicalPlan = pp, pp.Children()[0]
+		if sort, ok := dataSource.(*PhysicalSort); ok {
+			tail, dataSource = sort, sort.Children()[0]
+		}
```

Comment:
We can do an extra check here to make sure that dataSource`s NDV is larger than 1 if its histogram presents.
And if the NDV is larger than 1 but smaller than the concurrency, is it better to adjust the concurrency?

Reply:
OK~
I'm working on this.